common/network-common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java

```diff
@@ -155,10 +155,11 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
       // To avoid a race between TransportClientFactory.createClient() and this code which could
       // result in an inactive client being returned, this needs to run in a synchronized block.
       synchronized (this) {
+        boolean hasInFlightRequests = responseHandler.numOutstandingRequests() > 0;
```

Comment:
I think we should handle the synchronization when reading/writing with `responseHandler` to fix it. 
It seems that the `outstandingRequests` related variables of `responeHandler` is not guaranteed to be visible immediately once updated, and the update of `outstandingRequests` and `timeOfLastRequestNs` is not atomic, so that there are cases the fix here not working?